### DISCUSSION TITLE
Can't retrieve unregistered extension attribute 'language' when using nlp.pipe with n_process > 1 and spacy-langdetect

### DISCUSSION THREAD

**Discussion by mautini** [explosion/spaCy#9184] [Help: Coding & Implementations] [answered]
## Description of the problem
When I set n_process to 1 to a nlp.pipe after adding LanguageDetector to the pipeline, I got the following result :
```python
[{'language': 'en', 'score': 0.9999985260933938}]
```

but got 
```
AttributeError: [E046] Can't retrieve unregistered extension attribute 'language'. Did you forget to call the `set_extension` method?
```
when I set nb_process > 1 or -1

I thought it was due to the usage of spawn method (in multiprocessing) on MacOs since Python 3.8 and some context not sent to the subprocess but I got the same error on Linux using fork method.

## How to reproduce the behaviour
```python
import spacy
from spacy_langdetect import LanguageDetector
from spacy.language import Language


# Load Language detection
@Language.factory('language_detector')
def language_detector(nlp, name):
    return LanguageDetector()


# Load Spacy
nlp = spacy.load("en_core_web_lg")
nlp.add_pipe('language_detector', last=True)

print([doc._.language for doc in nlp.pipe(['I bless the rains down in Africa'], n_process=2)])
```

## Your Environment

* Operating System: MacOS 10.15 / Debian 11
* Python Version Used: 3.8.2 / 3.8.12
* spaCy Version Used: 3.1.2
* Environment Information: Spacy-langdetect 0.1.2

**Comment by adrianeboyd** [comment, marked as answer]
I was kind of surprised that the `spacy-langdetect` code worked at all. Usually we recommend defining the extension with `Doc.set_extension` globally rather than just on the `Doc` object with `doc.set_extension`.

A workaround for this code above is to call the component once before using multiprocessing:

```python
doc = nlp("test")
print([doc._.language for doc in nlp.pipe(['I bless the rains down in Africa'], n_process=2)])
```

`spacy-langdetect` is a third-party library that's not maintained by us, but you could try opening an issue in their repo. There's actually no particular need for this to be a pipeline component. You could just add the extensions globally with a utility function:

```python
import spacy
from spacy.tokens import Doc, Span, Token
from spacy_langdetect.spacy_langdetect import _detect_language

Doc.set_extension("language", getter=_detect_language)
Span.set_extension("language", getter=_detect_language)
Token.set_extension("language", getter=_detect_language)

print([doc._.language for doc in nlp.pipe(['I bless the rains down in Africa'], n_process=2)])
```

There's no need to set these repeatedly on each doc/token/span. You can just define them once and `nlp.pipe` should take care of the necessary context in multiprocessing.

**Reply by mautini** [reply]
Thanks, it works by adding the extension globally.

Unfortunately, `spacy-langdetect` seems to not be maintained anymore...

Thanks again for your time.
